mathesar/views/api.py

```diff
@@ -321,10 +321,14 @@ def list(self, request, table_pk=None):
 
     def retrieve(self, request, pk=None, table_pk=None):
         table = get_table_or_404(table_pk)
-        record = table.get_record(pk)
-        if not record:
-            raise NotFound
-        serializer = RecordSerializer(record)
+        try:
+            record = table.get_record(pk)
+            if not record:
+                raise NotFound
+            serializer = RecordSerializer(record)
+        except NotUniquePrimaryKey as e:
+            logger.error(e.message)
+            raise e
```

Comment:
You should be raising an `APIException` here, not the error directly. This will return the appropriate JSON error in the frontend. Please see how `APIException` is used elsewhere in the code. This also applies to other methods below.